### ISSUE TITLE
Provide multiple versions of documentation

### ISSUE THREAD

**Issue by kevinbarabash** [Khan/KaTeX#1509]
This is so that users can look up docs for the version that they're using even if it isn't the most recent version.

**Comment by ylemkimon** [comment]
Docusaurus supports documentation versioning: https://docusaurus.io/docs/en/versioning.

**Comment by kevinbarabash** [comment]
Awesome!

**Comment by ylemkimon** [comment]
https://github.com/Khan/KaTeX/issues/1505#issuecomment-408504458 by @edemaine 
> When we publish to gh-pages (which I guess we do every release, not every commit, so that we document the release not the master branch?), 

Another question is should we document the `master` branch, i.e., publish the website(`gh-pages`) every commit?

**Comment by ylemkimon** [comment]
I think there can be two workflows:
1. Publish website every commit to `master` and every release
* Use CircleCI to build and deploy the website using GitHub token or SSH deploy key
* Run `deploy` job on `master` branch commit builds and Git tag build
* Commit and use local built CSS and fonts or use CircleCI build artifacts for `master` documentation
  * **CORS is not enabled for CircleCI build artifacts, so it needs a proxy**
* Use CSS and fonts from CDN for release version documentation
* Latest release version documentation is shown by default
* `master` documentation can be accessed by selecting `next`
* **May bloat the repo**

2. Publish website every release
* Publish the website manually in the release script
* Use CSS and fonts from CDN

Or we can switch to `netlify`, which provides with following limitations:

> Network Bandwidth: 100GB/month - Soft
Storage: 100GB - Soft
API requests: 500 requests/minute, 3 deploys/minute* - Hard

* Use netlify to build and deploy the website
* Use local built CSS and fonts
* Use CSS and fonts from CDN for release version documentation
* Latest release version documentation is shown by default
* `master` documentation can be accessed by selecting `next`
* PR preview is available through #1510

**Comment by ronkok** [comment]
> Publish website every release

The function support page has been updated on that schedule for 1½ years. It works fine. I'm not sure about the other pages. We may find a problem and wish to get new docs out in between releases.

**Comment by edemaine** [comment]
I think we're currently in Plan 2 (though adding to release script sounds nice), but with this issue completed, Plan 1 makes sense. This is why I was thinking it might be good for the published compiled documentation (not the doc source) to live in a separate repo, to avoid any bloat.

A small issue regarding the function support page: in the past, @ronkok has done the work immediately *after* a release. If we continue this workflow, we'll need to be able to update a past release's documentation. (This is also an issue if we want to publish the docs on NPM.) The alternative would be to prepare the doc update before release.

**Comment by ylemkimon** [comment]
@edemaine I think documentations should be updated in the PR that adds a new feature.

**Comment by ronkok** [comment]
> Perhaps we should start requiring updates to the document for each PR that affects it.

That may be more easily said than done. The original KaTeX Contributing page instructed contributors to update the wiki's function support page. But I found in December 2015 that the function support page was missing several hundred functions. And so began my involvement here.

> in the past, @ronkok has done the work immediately after a release.

Give me a few days notice before a release occurs and I can write up every function added since the previous release.  I expect the velocity of new functions to decrease.

Now I certainly would not mind if others updated the function support page. But that is less easy to do now than it was in the wiki days, and even then, people quite often didn't do it. Also, a requirement like that may discourage new contributors just by adding another burden.

I also have reservations about showing functions in the documentation that are not yet released. They should at least be flagged as such. Which means that the function support page must then be edited twice for each new function, once with the flag, and once to remove the flag.

If you can get contributors to do all that, great. If not, I'll follow up and keep the page current.

**Comment by edemaine** [comment]
@ronkok All good points. We already generally require people to write tests for new features, so I think requiring documentation writing to be good practice. We can modify the contributing document, but also PR templates (essentially #1468) would help here: we could add checkboxes like

- [ ] Tests for new functionality or checking for old bug
- [ ] New functionality documented in `docs`, if applicable

Nonetheless, some mistakes will creep through (`cli.js` has often fallen behind on settings, for example), and I like the idea of a review of the new features and possibly a cleaning up of documentation before each release.  I suggest a process of something like: open an issue "Preparing for 0.11.0", and everyone who wants to can review what they care about, probably especially as regards documentation, but also things like "I'd like to finish PR #n first".  Maybe the issue includes a draft of the release notes.

I definitely agree that the master documentation (`next`) should not be the default view.  But it could be nice to have things there as an option, especially given that we're currently pretty slow to release.  (Not an ideal state, but the truth for now.)  I think we can add a general warning in the `next` documentation (probably Docusaurus can do this for us) that says "Warning: these features are available only on the master branch of Github, and are subject to change."

**Comment by edemaine** [comment]
(moving over from https://github.com/Khan/KaTeX/pull/1520#issuecomment-409001490)
I'd like to reraise the question of whether we should have a separate repo for the documentation/website publication (not the source files).  We could move the website to `http://katex.github.io/` which would be served from a repository called `katex.github.io` within the KaTeX project.  (See https://help.github.com/articles/user-organization-and-project-pages/#user-and-organization-pages-sites)  Then we wouldn't have to worry about repo size, and we would be free to put all the versions of documentation there (which really wants to have the CSS/JS/fonts copied).

**Comment by ylemkimon** [comment]
It seems it comes down to three questions:

**1. Where we should we publish the website?**
* KaTeX repo GitHub Pages, `gh-pages` branch (https://khan.github.io/KaTeX/)
* KaTeX organization GitHub Pages (https://KaTeX.github.io)
* netlify (https://katex.netlify.com/)

**2. If we choose to use GitHub Pages, should we publish `master`(next) documentation to GitHub Pages? (Currently, it is published to netlify) Or should we have a link to netlify?**

**3. If we choose to publish `master` documentation to GitHub Pages, where should we serve KaTeX from?**
* Commit to GitHub Pages
* CircleCI build artifacts
* netlify

**Comment by ylemkimon** [comment]
@edemaine 
> `gh-pages` means possible repo bloat, so should probably aim for CircleCI build artifacts. I'm not an expert on how easy this is to do. Two questions: for released versions, and for master branch?

For released versions, we can use KaTeX from CDN.
For `master` branch, CircleCI doesn't have CORS header set, i.e., doesn't allow accessing the artifacts from another site. Therefore, we should use (CORS) proxy, but I'm not sure there is free and reliable one. So probably we will have to use another object storage like AWS S3.


> KaTeX organization GitHub Pages avoids the bloat issue, so we can do whatever we want (so it's my leaning).

KaTeX organization GitHub Pages is still a GitHub repo, which has size limits. I guess we can prune old histories regularly. I wonder how quickly the repo bloats.

> netlify has its limits on network bandwidth etc. that I'm slightly worried about, but maybe shouldn't be. 

It turns out that GitHub Pages also has similar limitations:

Usage | GitHub Pages | netlify
---- | -------- | --------
Storage | 1GB (hard, recommended 1GB for repo) | 100GB (soft, total)
Network Bandwidth | 100GB/month (soft) | 100GB/month (soft)
Build | 10 builds/hour(soft) | 180 builds/hour(hard)

**Comment by kevinbarabash** [comment]
I think before we make any change to where the site lives it would be nice if we could secure a stable domain name for katex and switch to that first before changing where files are served from.  Also, since people are worried about bloat let's hold off on having multiple versions of the docs until we have a stable domain.

**Comment by ylemkimon** [comment]
@kevinbarabash 
> I think before we make any change to where the site lives it would be nice if we could secure a stable domain name for katex and switch to that first before changing where files are served from.

Some ideas:
* katex.io
* katexjs.org

> Also, since people are worried about bloat let's hold off on having multiple versions of the docs until we have a stable domain.

Versioning (#1521) itself will not bloat the repo as it uses KaTeX from CDN.

**Comment by edemaine** [comment]
I can buy those domains (and possibly transfer ownership) if that would speed things up here (but also don't want to complicate things).  @kevinbarabash What's best from KA's perspective?

**Comment by kevinbarabash** [comment]
I can ask our IT dept about buying a domain.

**Comment by ylemkimon** [comment]
`js.org` seems to [provide free subdomains for GitHub pages](https://github.com/js-org/js.org). So if we decide to publish at GitHub pages, another option is `katex.js.org`.

**Comment by edemaine** [comment]
Nice find! `katex.js.org` is another reasonable choice, especially if KA doesn't want to buy one.

**Comment by ylemkimon** [comment]
I think we can use CircleCI build artifacts using CORS proxy on `now.sh` and `Cloudflare` to cache/deliver files. I've created a demo at https://katex.ylem.kim and build artifacts are available in https://katex.ylem.kim/:build_num/:path.

**Comment by ylemkimon** [comment]
I've opened several PRs, as I think it'd be easier to understand and decide with actual codes.

If we publish the main website to netlify, we can use local built CSS as it provides 100 GB storage. (#1584 [1])

If we publish the main website to GitHub pages (`gh-pages` branch or organization repo), we can commit local built CSS. (#1584 [1])

But this may bloat the repo, so we can remove master (`next`) documentation from GitHub pages. (#1585 [2-1]) And maybe we can link to netlify (https://katex.netlify.com/).

Alternatively, we can serve KaTeX from another source. (#1586 [2-2])

**Comment by ylemkimon** [comment]
Another question is should we use release (stable) version or master (latest) version for the main page.

**Comment by ronkok** [comment]
I think the main page should get the release (stable) version. Otherwise we can count on seeing issues raised asking, "Why doesn't my page render a ____?".

**Comment by edemaine** [comment]
I think the simplest starting point, roughly where we are now, is to publish documentation only during release, and use CDN for JS/CSS/fonts.  We should do what's necessary for that first.  I believe #1584 is good for that, too, because it enables local testing.

To me, the main options are:
* Publish to GitHub pages in organization repo. Ideally linked from a domain name, though I also think `katex.github.io` is also a fine URL (better than the current one).  [my preference]
* Publish to Netlify [1]. Main downside is reliance on another tool (whereas our current reliance on netlify is nonessential, just for testing). Also needs a domain name.
* We can also do a hybrid that uses GitHub pages in `gh-pages` but stores the built master files elsewhere (Netlify is probably the neatest idea there), but these seem overly complicated.

**Comment by ylemkimon** [comment]
> We can also do a hybrid that uses GitHub pages in gh-pages but stores the built master files elsewhere (Netlify is probably the neatest idea there), but these seem overly complicated.

I'm thinking of adding the ability to run previous versions or test a PR on the REPL, so it is eventually needed.

**Comment by kevinbarabash** [comment]
I think we should wait on getting a domain name that is independent of which tool we're using before making changes to where the docs live.

**Comment by kevinbarabash** [comment]
Do you have to serve files from github.com to use js.org?

**Comment by ylemkimon** [comment]
Now I'm against using `js.org` as we don't have control over DNS and subdomains.

**Comment by kevinbarabash** [comment]
Not having control over DNS would limit us to keeping files in Khan/KaTeX or KaTeX/katex.github.io, right?  @ylemkimon how were you thinking we'd use subdomains?

**Comment by ylemkimon** [comment]
@kevinbarabash Yes, it would limit us to GitHub pages or only production deploy of netlify. netlify uses subdomain for deploy previews, e.g., https://deploy-preview-1569--katex.netlify.com. Also if we choose to publish only release docs on the main website, we can serve latest docs on the `next.katex.io`, for example.

**Comment by edemaine** [comment]
@rustylife You seem to own `katex.org` (apologies if you just happen to have the same name as the WHOIS entry). Would you consider transferring ownership? Please email me if so!

**Comment by ylemkimon** [comment]
Now we've secured domain, restating the question:

**Where we should we publish the (main) website?**
* KaTeX repo GitHub Pages, `gh-pages` branch (https://khan.github.io/KaTeX/)
* KaTeX organization GitHub Pages (https://KaTeX.github.io)
* netlify (https://katex.netlify.com/)

**Comment by edemaine** [comment]
@rustylife That's great! Could you unlock the domain (if locked) and email me an authorization code / EPP code / transfer code?  Email to edemaine@mit.edu, not this public thread.

**Comment by kevinbarabash** [comment]
@rustylife and @edemaine thank you both.

**Comment by ylemkimon** [comment]
@edemaine We need a DNS to point to the website. If we decide to use Netlify, we can use Netlify DNS. If we use GitHub Pages, I think we can use Cloudflare.

**Comment by edemaine** [comment]
@ylemkimon Just to add to the list of options, we can also use Namecheap's free DNS, given that I'm moving it to Namecheap (at least for the duration that I hold it, before it transfers to KA). I can give you (and others) access to that DNS system once it's moved. But I'm happy to use anything else. I guess Netlify makes sense if we want it to manage our subdomains.

**Comment by ylemkimon** [comment]
@edemaine Unfortunately, domain registrar's DNSes usually don't have uptime guarantees, many POPs, and an option for setting root domain CNAME (e.g., ALIAS/ANAME record, CNAME flattening). Cloudflare also has a free plan and it provides CDN (caching).

**Comment by ylemkimon** [comment]
It seems the domain has successfully been transferred.

**Comment by ylemkimon** [comment]
Now I think we can close this issue now. Thank you all for the great work!
